### PR TITLE
add namespace to the resourcereference

### PR DESCRIPTION
**Pull request by gaoning777**
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/pipelines/2651)

### CODE REVIEW

File: backend/api/run.proto [hunk 1]
```diff
@@ -188,6 +191,7 @@ message UnarchiveRunRequest {
 
 message DeleteRunRequest {
   string id = 1;
+  string namespace = 2;
```
Comment (Ark-kun) [inline backend/api/run.proto]: Looks like the run can be uniquely identified by the run ID, so the namespace might be unnecessary. Same with other API methods that work on existing runs.

**Comment by Ark-kun** [comment]
Is "namespace" the only user isolation implementation we have in mind for the future?
Should we use something generic like "user" or "isolation_group" in the API **surface**?

File: backend/src/apiserver/resource/model_converter.go [hunk 2]
```diff
@@ -199,15 +199,17 @@ func (r *ResourceManager) toModelResourceReferences(
...
 			return nil, util.Wrap(err, "Failed to find the referred resource")
 		}
-		modelRef := &model.ResourceReference{
-			ResourceUUID:  resourceId,
-			ResourceType:  resourceType,
-			ReferenceUUID: apiRef.Key.Id,
-			ReferenceName: referenceName,
-			ReferenceType: modelReferenceType,
-			Relationship:  modelRelationship,
+		if apiRef.Key.Type != api.ResourceType_NAMESPACE {
```
Comment (Bobgy) [inline backend/src/apiserver/resource/model_converter.go]: Just curious, why is namespace skipped here?

File: backend/src/apiserver/common/const.go [hunk 3]
```diff
@@ -28,11 +28,13 @@ const (
 	Run             ResourceType = "Run"
 	Pipeline        ResourceType = "pipeline"
 	PipelineVersion ResourceType = "PipelineVersion"
+	Namespace				ResourceType = "Namespace"
```
[the remark this answers is not in the document]
Comment (gaoning777) [reply backend/src/apiserver/common/const.go]: I've used the gofmt to format the go codes. Thanks

File: backend/src/apiserver/resource/model_converter.go (hunk 2 above)
Comment (gaoning777) [reply backend/src/apiserver/resource/model_converter.go]: This function is called when the backend tries to construct the data structure for the db. Currently, we haven't intended to add the namespace field in the db yet. We will in the near future.

File: backend/api/resource_reference.proto [hunk 4]
```diff
@@ -23,12 +23,14 @@ enum ResourceType {
...
   PIPELINE = 3;
   PIPELINE_VERSION = 4;
+  NAMESPACE = 5;
 }
 
 enum Relationship {
   UNKNOWN_RELATIONSHIP = 0;
   OWNER = 1;
   CREATOR = 2;
+  BELONGING = 3;
```
Comment (IronPan) [inline backend/api/resource_reference.proto]: Does it make sense to model namespace as the owner of the other resources?
The owner can potentially trigger the cascade delete in delete API call.

File: backend/src/apiserver/resource/model_converter.go [hunk 5]
```diff
@@ -244,6 +246,8 @@ func (r *ResourceManager) getResourceName(resourceType common.ResourceType, reso
 			return "", util.Wrap(err, "Referred pipeline version not found.")
 		}
 		return version.Name, nil
+	case common.Namespace:
```
Comment (IronPan) [inline backend/src/apiserver/resource/model_converter.go]: What would happen if user delete the namespace and later someone else create a namespace with the same name? Would it cause the data leak?
Should we store the uuid of a namespace instead? (this is not super urgent but we might want to think through)

File: backend/api/resource_reference.proto (hunk 4 above)
Comment (gaoning777) [reply backend/api/resource_reference.proto]: Sure. I can use the OWNER relationship.

File: backend/src/apiserver/resource/model_converter.go (hunk 5 above)
Comment (gaoning777) [reply backend/src/apiserver/resource/model_converter.go]: That is a very good point. But, could it be that reusing the same name is the intended behavior? I will put a note here for further investigation.

**Pull request merged**